### PR TITLE
Make the search block editor match the frontend

### PR DESCRIPTION
**Pull request by youknowriad**
Related #29976

This block also relies on "input" and "button" element which in wp-admin has some opinionated styles, this PR resets those styles in the editor canvas.

This PR needs to be tested properly (specially the editor UI to ensure the select reset is not affecting anything that shouldn't).

At the moment, this PR impact placeholders in other blocks because these resets have higher specificity than the UI components used in placeholders (Button, TextControl...). We have three options:

 - ~Use a similar technique as  in #29229 and add `:not(.components-button)` and similar to the resets.~
 - iFrame the placeholders or use shadow DOM for them because their content shouldn't be inheriting the styles from the canvas.
 - Add more specific styles to the components (button, text control...) when used inside placeholders.

### CODE REVIEW

File: packages/block-library/src/search/edit.js [hunk 1]
```diff
@@ -148,16 +149,21 @@ export default function SearchEdit( {
...
-						withoutInteractiveFormatting
-						value={ buttonText }
-						onChange={ ( html ) =>
-							setAttributes( { buttonText: html } )
-						}
-					/>
+					// This rule only makes sense for UI not blocks.
+					// eslint-disable-next-line react/forbid-elements
+					<button type="submit" className="wp-block-search__button">
+						<RichText
```
Comment (youknowriad) [inline packages/block-library/src/search/edit.js:155]: It seems there's something that prevents RichText from working as a "button". I've used a small hack here to make it work and still keep a similar markup to the frontend.

Comment (ellatrix) [reply packages/block-library/src/search/edit.js:155]: What prevents it from working? Do you get any errors?

Comment (youknowriad) [reply packages/block-library/src/search/edit.js:155]: It's just stuck, like you can't edit it. Maybe "contenteditable" doesn't work on buttons?

Comment (ellatrix) [reply packages/block-library/src/search/edit.js:155]: That's interesting. I'll try this out in a bit.

Comment (ellatrix) [reply packages/block-library/src/search/edit.js:155]: It should work: https://codepen.io/iseulde/pen/vYgOpVw

Maybe something is wrong in RichText.

Comment (ellatrix) [reply packages/block-library/src/search/edit.js:155]: For some reason the selection change event is not firing in this context. It is in codepen though. Strange.

Comment (ellatrix) [reply packages/block-library/src/search/edit.js:155]: It seems something is the block editor is causing the issue. I insert `<button contenteditable>test</button>` without rich text in a block and it also fails to set selection. The same line on an empty web page works fine.

File: packages/block-library/src/reset.scss [hunk 2]
```diff
@@ -98,4 +98,40 @@
 		line-height: revert;
 		font-weight: revert;
 	}
+
+	select,
+	input,
+	textarea,
+	button {
+		font-family: system-ui;
+		all: revert;
```
Comment (ellatrix) [inline packages/block-library/src/reset.scss]: Just for the public record :) `all: revert;` is also reverting [`-webkit-user-modify`](https://developer.mozilla.org/en-US/docs/Web/CSS/user-modify) to `read-only` preventing `contenteditable` from working.

**Comment by carolinan** [comment]
Empty theme, editor:
the search block styles in the editor, with the pr applied

Front:
the search block styles on the front

Broken styles in Twenty Twenty One, front view
Before:
search block variations in Twenty Twenty one, without the PR applied

After:
The search block in Twenty Twenty one is missing border styles when the PR is applied


As with the select list, I would like the defaults to be more usable, they need to be good enough to use _as is_ without forcing theme and plugin developers to add their custom styles. 
I feel like at the very least, the height of and the spacing in the input field should be the same no matter where the button and  icon is placed.
_TLDR: Make it match, but make it pretty._

**Comment by youknowriad** [comment]
> TLDR: Make it match, but make it pretty.

I agree with this but I don't think it should be done in these PRs because it has implications and impact on themes which can be higher and it should be done very carefully. For now, the purpose is just to make the backend look like the frontend without impacting the frontend. I'll fix 2021

File: packages/block-library/src/search/edit.js (hunk 1 above)
Comment (youknowriad) [reply packages/block-library/src/search/edit.js:155]: @ellatrix after the removal of revert: all, I have two bugs:

 - In safari, backspace removes the block and not a character in the text
 - in chrome, I can't type an empty space

Comment (ellatrix) [reply packages/block-library/src/search/edit.js:155]: Hm, yes, Space will conflict with "pressing" the button.
I'll have a look at both issues. In the meantime, feel free to nest it in a span and we can adjust later.

**Comment by jasmussen** [comment]
I really appreciate this PR. I just ran into issues where the use of editor components adds some extra classes and therefore specificity to the rules, which means it's really really tricky for a theme to style the form widgets of this one in an elegant way. It's specifically the `components-button` was problematic, so that change alone is a big step forward. And in TT1 blocks I see this before, in the editor:

Screenshot 2021-04-26 at 09 21 54

and on the frontend:

Screenshot 2021-04-26 at 09 22 01

And here's after, in the editor:

Screenshot 2021-04-26 at 09 24 30

and on the frontend:

Screenshot 2021-04-26 at 09 24 36

In other words, to me this solves a problem of the style inheritance very nicely.

**Comment by youknowriad** [comment]
Yes, this is a good PR. Unfortunately, the reset used here conflicts with placeholder buttons. It's the same issue we have with themes styling `button`. So it's kind of blocks by this.

**Comment by youknowriad** [comment]
Closing this as still blocked by CSS bleed in placeholder components. It's unclear how we'll solve this.
